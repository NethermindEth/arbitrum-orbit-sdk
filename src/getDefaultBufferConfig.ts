import { Client, Transport, Chain } from 'viem';

import { ParentChainId, validateParentChain } from './types/ParentChain';
import { getParentChainBlockTime } from './getParentChainBlockTime';

// todo: get from abi
export type BufferConfig = {
  threshold: bigint;
  max: bigint;
  replenishRateInBasis: bigint;
};

export function getDefaultBufferConfig<TChain extends Chain | undefined>(
  parentChainIdOrClient: ParentChainId | Client<Transport, TChain>,
): BufferConfig {
  const { chainId: parentChainId, isCustom: parentChainIsCustom } =
    validateParentChain(parentChainIdOrClient);

  if (parentChainIsCustom) {
    throw new Error(
      `[getDefaultBufferConfig] can't provide defaults for custom parent chain with id ${parentChainId}`,
    );
  }

  const blocksPerMinute = 60 / getParentChainBlockTime(parentChainId);

  return {
    // 2 hours
    threshold: BigInt(2 * 60 * blocksPerMinute),
    // 2 days
    max: BigInt(2 * 24 * 60 * blocksPerMinute),
    // 5%
    replenishRateInBasis: BigInt(500),
  };
}
