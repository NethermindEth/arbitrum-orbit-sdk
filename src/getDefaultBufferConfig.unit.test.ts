import { it, expect } from 'vitest';

import { chains } from './chains';
import { ParentChainId } from './types/ParentChain';
import { getDefaultBufferConfig } from './getDefaultBufferConfig';

it('returns default value for bufferConfig based on parent chain', () => {
  expect(
    chains
      .filter((chain) => chain.id !== 333333)
      .reduce((acc, value) => {
        return {
          ...acc,
          // it's ok to cast as we've filtered out 333333 above
          [value.id]: getDefaultBufferConfig(value.id as ParentChainId),
        };
      }, {}),
  ).toMatchSnapshot();
});
