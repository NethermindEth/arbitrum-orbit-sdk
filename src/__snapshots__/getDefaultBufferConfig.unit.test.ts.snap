// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`returns default value for bufferConfig based on parent chain 1`] = `
{
  "1": {
    "max": 14400n,
    "replenishRateInBasis": 500n,
    "threshold": 600n,
  },
  "11155111": {
    "max": 14400n,
    "replenishRateInBasis": 500n,
    "threshold": 600n,
  },
  "1337": {
    "max": 14400n,
    "replenishRateInBasis": 500n,
    "threshold": 600n,
  },
  "412346": {
    "max": 14400n,
    "replenishRateInBasis": 500n,
    "threshold": 600n,
  },
  "42161": {
    "max": 14400n,
    "replenishRateInBasis": 500n,
    "threshold": 600n,
  },
  "421614": {
    "max": 14400n,
    "replenishRateInBasis": 500n,
    "threshold": 600n,
  },
  "42170": {
    "max": 14400n,
    "replenishRateInBasis": 500n,
    "threshold": 600n,
  },
  "8453": {
    "max": 86400n,
    "replenishRateInBasis": 500n,
    "threshold": 3600n,
  },
  "84532": {
    "max": 86400n,
    "replenishRateInBasis": 500n,
    "threshold": 3600n,
  },
}
`;
